addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandlerImpl.java

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.modbus.handler;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.modbus.ModbusBindingConstants;
+import org.openhab.binding.modbus.internal.config.ModbusPollerConfiguration;
+import org.openhab.io.transport.modbus.BitArray;
+import org.openhab.io.transport.modbus.ModbusManager;
+import org.openhab.io.transport.modbus.ModbusManager.PollTask;
+import org.openhab.io.transport.modbus.ModbusReadCallback;
+import org.openhab.io.transport.modbus.ModbusReadFunctionCode;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprintImpl;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.openhab.io.transport.modbus.PollTaskImpl;
+import org.openhab.io.transport.modbus.endpoint.ModbusSlaveEndpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ModbusPollerThingHandlerImpl} is responsible for polling Modbus slaves. Errors and data is delegated to
+ * child thing handlers inheriting from {@link ModbusReadCallback} -- in practice: {@link ModbusDataThingHandler}.
+ *
+ * @author Sami Salonen - Initial contribution
+ */
+public class ModbusPollerThingHandlerImpl extends BaseBridgeHandler implements ModbusPollerThingHandler {
+
+    /**
+     * {@link ModbusReadCallback} that delegates all tasks forward.
+     *
+     * All instances of {@linkplain ReadCallbackDelegator} are considered equal, if they are connected to the same
+     * bridge. This makes sense, as the callback delegates
+     * to all child things of this bridge.
+     *
+     * @author Sami Salonen
+     *
+     */
+    private class ReadCallbackDelegator implements ModbusReadCallback {
+
+        private void forEachAllChildCallbacks(Consumer<ModbusReadCallback> callback) {
+            getThing().getThings().stream()
```

Comment:
`Poller` thing is responsible for polling the data and is receiving the updates. It's duty is to propagate the updates to handlers of child things.

Would it be OK that poller thing keeps track of child handlers using `childHandlerInitialized` and `childHandlerDisposed`? This way there would not be any full iteration.

Certainly it is also possible for child handlers to "register" themselves for data updates, but I don't see the benefit adding this kind of registration mechanism. After all, currently thing hierarchy is already conveying the same thing, telling which data thing is interested in which pollers' data.

Reply:
I have thought about your concept again and why you have the 3 layers: TCP-Bridge -> Poller-Bridge -> Data-Thing. For your use-case it certainly makes sense to have these 3 because of the limitation of the underlying modbus protocol. Since you say that one should not register too many "readers" on the bus for certain types of modbus-requests the "poller bridge" can serve as a router for multiple Data-Things.

You suggest using `childHandlerInitialized` and `childHandlerDisposed`, which I think is a better implementation than iterating blindly over the things. Since the whole purpose of a Data-Thing is to consume data which is retrieved by it's bridge, I think its OK to not use a registration and notification mechanism.